src/cloudevents/core/formats/base.py

```diff
@@ -0,0 +1,24 @@
+#  Copyright 2018-Present The CloudEvents Authors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+from typing import Protocol, Union
+
+from cloudevents.core.base import BaseCloudEvent
+
+
+class Format(Protocol):
+    def read(self, data: Union[str, bytes]) -> BaseCloudEvent: ...
+
+    def write(self, event: BaseCloudEvent) -> str: ...
```

Comment:
So in order to make this useful as a generic approach we should add generics :-)

```suggestion
T = TypeVar("T", bound=BaseCloudEvent)

class Format(Protocol[T]):
    def read(self, data: bytes) -> T: ...

    def write(self, event: T) -> bytes: ...
```


this way we should be able to get parameterized results as well

Reply:
good idea, yes, thanks, I'll do that